features/steps/thens/geometry.py

```diff
@@ -36,79 +38,112 @@ def step_impl(context, inst):
+            bounds = [vs[edges_idxs]]
+        else:
+            bounds = []
+            for bnd in indices:
+                edges_idxs = numpy.array(list(zip(bnd, numpy.roll(bnd, -1))))
+                edges = vs[edges_idxs]
+                bounds.append(edges)
     else:
-        loop = ifcopenshell.geom.create_shape(settings, inst)
-        verts = numpy.array(loop.verts).reshape((-1, 3))
-        edge_idxs = numpy.array(loop.edges).reshape((-1, 2))
-        edges = verts[edge_idxs]
+        # We set deflection (max deviation from real to approximated) based on the precision
+        # it needs to be significantly larger than the precision value, because otherwise we
+        # can get false positive intersection outcomes (near sharp corners). But also not too
+        # large because that can also result into false positives (when multiple curves are
+        # close)
+        deflection = min(precision * 10, 0.01)
+        try:
+            settings = ifcopenshell.geom.settings(MESHER_LINEAR_DEFLECTION=deflection)
+        except:
+            settings = ifcopenshell.geom.settings()
+            settings.set_deflection_tolerance(deflection)
 
-    for i, edge in enumerate(edges):
-        if isinstance(edge, numpy.ndarray):
-            ps = edge
+        if USE_MAPPING:
+            loop = ifcopenshell.ifcopenshell_wrapper.map_shape(ifcopenshell.geom.settings(), inst.wrapped_data)
+            bounds = [list(loop.children)]
         else:
-            if edge.basis:
-                raise NotImplementedError()
-            ps = numpy.array([
-                edge.start.coords,
-                edge.end.coords
-            ])
-        idx.insert(i, (ps.min(axis=0) - precision).tolist() + (ps.max(axis=0) + precision).tolist())
+            loop = ifcopenshell.geom.create_shape(settings, inst)
+            verts = numpy.array(loop.verts).reshape((-1, 3))
+            edge_idxs = numpy.array(loop.edges).reshape((-1, 2))
+            bounds = [verts[edge_idxs]]
 
-    for i, edge in enumerate(edges):
-        if isinstance(edge, numpy.ndarray):
-            ps = edge
-        else:
-            ps = numpy.array([
-                edge.start.coords,
-                edge.end.coords
-            ])
-        ps_flat = numpy.concatenate((ps.min(axis=0),ps.max(axis=0)))
-        neighbours = {(i - 1) % len(edges), (i + 1) % len(edges)}
-        for j in idx.intersection(ps_flat):
-            if i <= j:
-                continue
-            if isinstance(edges[j], numpy.ndarray):
-                qs = edges[j]
+    for edges in bounds:
+        for i, edge in enumerate(edges):
+            if isinstance(edge, numpy.ndarray):
+                ps = edge
             else:
-                qs = numpy.array([
-                    edges[j].start.coords,
-                    edges[j].end.coords
+                if edge.basis:
+                    raise NotImplementedError()
```

Comment:
I seem to remember from earlier discussions that raising exceptions was problematic in terms of properly tracking the validation outcome and severity.